### ISSUE TITLE
ServerStream SendAndClose Does Not Close Stream

### ISSUE THREAD

**Issue by notbdu** [grpc/grpc-go#1755]
Please answer these questions before submitting your issue.

### What version of gRPC are you using?
`v1.8.1`

### What version of Go are you using (`go version`)?
`1.8.1`

### What operating system (Linux, Windows, …) and version?
```
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.2 LTS
Release:        16.04
Codename:       xenial
```

### What did you do?
For a unidirectional grpc streams endpoint, calling the autogenerated grpc server stream method `SendAndClose` does not close the stream. The client continues to be able to `Send` over the stream even though the server method handler has already returned and completed.

Service method definition:
```
rpc SomeMethod(stream Request)
        returns (Response) {}
```

Autogenerated `SendAndClose` method:
```
func (x *someMethodServer) SendAndClose(m *Response) error {
        return x.ServerStream.SendMsg(m)
}
```

User implemented service method handler function:
```
func (s *server) SomeMethod(stream Service_SomeMethodServer) error {
    defer stream.SendAndClose(&Response{})
    ...
    return nil
}
```

Client code calling send:
```
if err := stream.Send(&Request{}); err != nil {
        return
}
```

### What did you expect to see?
Calling `stream.SendAndClose()` should close the stream and clients attempting to call `stream.Send()` should receive an error.

### What did you see instead?
The client code continues to `Send` over the `stream` w/o `err` even though the server `SomeMethod` handler has returned and called `stream.SendAndClose()`.

It looks like the autogenerated `SendAndClose` method just calls `SendMsg` w/o explicitly closing the stream. This could work client side if the client code could block on `stream.Recv()` and exit when a response is received. But since this is a unidirectional stream, client streams do not have a `Recv()` method.

**Comment by dfawley** [comment]
IMO `SendAndClose` is poorly named, and the server handler generated for client-streaming RPCs should have _returned_ the response message and error like unary handlers, but the API is unfortunately not something we are able to change at this point.

It looks like there is a bug here preventing client-side `Send`s from returning an error when the stream is closed.  We'll look into it.  Note that the same thing also appears to be true when the client calls `CloseSend` and then performs subsequent `Send` calls.

As a minor side note, I would not recommend deferring `stream.SendAndClose`; instead `return` either an `error` or `stream.SendAndClose(response)` (which itself returns an `error` if a problem is encountered) from your handler.  Otherwise, if there is an error, your server will transmit the response message to the client, even though it will be ignored (due to the code in `CloseAndRecv`).

**Comment by notbdu** [comment]
Thanks @dfawley for the quick response and feedback.

**Comment by dfawley** [comment]
I did some digging, and it looks like an error will eventually be returned from `Send`, but only after the flow control limit is reached, which could be as much as 4MB.

Like #1738, there will be a lot of changes going into the transport Real Soon Now, so we will wait until those are in before putting in a fix for this.

**Comment by dfawley** [comment]
This should also be fixed by #1792 (the same root cause).

**Comment by josephusmv** [comment]
hi, dfawley, 

very glad to see there are go grpc developers agree that SendAndClose is poorly named. And I think this lead to some confusion in my application level codes and engineers who read my codes complain about it. So I opened an information issue #2156 to ask about whether you have any plan to do some compensation about it. Please have a look.

thanks
